File: frontend/src/api/local/diseases.ts
/**
 * Local Diseases/Health Tracking API
 */

import type {
  DiseaseRecord, DiseaseRecordCreate, DiseaseRecordUpdate,
  DiseaseRecordDetail, DiseaseTreatment, DiseaseTreatmentCreate,
  DiseaseHealthSummary,
} from '../../types'
import { db, generateId, now, getLocalUserId } from './helpers'

function rowToRecord(row: any): DiseaseRecord {
  return {
    id: row.id,
    livestock_id: row.livestock_id,
    tank_id: row.tank_id,
    user_id: row.user_id,
    disease_name: row.disease_name,
    symptoms: row.symptoms || null,
    diagnosis: row.diagnosis || null,
    severity: row.severity,
    status: row.status,
    detected_date: row.detected_date,
    resolved_date: row.resolved_date || null,
    outcome: row.outcome || null,
    notes: row.notes || null,
    created_at: row.created_at,
    updated_at: row.updated_at,
  }
}

function rowToTreatment(row: any): DiseaseTreatment {
  return {
    id: row.id,
    disease_record_id: row.disease_record_id,
    user_id: row.user_id,
    consumable_id: row.consumable_id || null,
    treatment_type: row.treatment_type,
    treatment_name: row.treatment_name,
    dosage: row.dosage || null,
    quantity_used: row.quantity_used ?? null,
    quantity_unit: row.quantity_unit || null,
    treatment_date: row.treatment_date,
    duration_days: row.duration_days ?? null,
    effectiveness: row.effectiveness || null,
    notes: row.notes || null,
    created_at: row.created_at,
  }
}

export const diseasesApi = {
  list: async (params?: {
    tank_id?: string; livestock_id?: string; status?: string; severity?: string
  }): Promise<DiseaseRecord[]> => {
    const userId = getLocalUserId()
    const conditions = ['user_id = ?']
    const values: any[] = [userId]

    if (params?.tank_id) { conditions.push('tank_id = ?'); values.push(params.tank_id) }
    if (params?.livestock_id) { conditions.push('livestock_id = ?'); values.push(params.livestock_id) }
    if (params?.status) { conditions.push('status = ?'); values.push(params.status) }
    if (params?.severity) { conditions.push('severity = ?'); values.push(params.severity) }

    const rows = await db.query(
      `SELECT * FROM disease_records WHERE ${conditions.join(' AND ')} ORDER BY detected_date DESC`,
      values
    )
    return rows.map(rowToRecord)
  },

  get: async (id: string): Promise<DiseaseRecordDetail> => {
    const row = await db.queryOne('SELECT * FROM disease_records WHERE id = ?', [id])
    if (!row) throw new Error('Disease record not found')
    const record = rowToRecord(row)

    const treatmentRows = await db.query(
      'SELECT * FROM disease_treatments WHERE disease_record_id = ? ORDER BY treatment_date DESC',
      [id]
    )

    return { ...record, treatments: treatmentRows.map(rowToTreatment) }
  },

  create: async (data: DiseaseRecordCreate): Promise<DiseaseRecord> => {
    const id = generateId()
    const userId = getLocalUserId()
    const timestamp = now()

    await db.execute(
      `INSERT INTO disease_records (id, livestock_id, tank_id, user_id, disease_name, symptoms, diagnosis,
       severity, status, detected_date, resolved_date, outcome, notes, created_at, updated_at)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      [id, data.livestock_id, data.tank_id, userId, data.disease_name,
       data.symptoms ?? null, data.diagnosis ?? null,
       data.severity ?? 'moderate', data.status ?? 'active',
       data.detected_date, data.resolved_date ?? null,
       data.outcome ?? null, data.notes ?? null, timestamp, timestamp]
    )

    const row = await db.queryOne('SELECT * FROM disease_records WHERE id = ?', [id])
    return rowToRecord(row)
  },

  update: async (id: string, data: DiseaseRecordUpdate): Promise<DiseaseRecord> => {
    const fields = ['disease_name', 'symptoms', 'diagnosis', 'severity', 'status',
      'detected_date', 'resolved_date', 'outcome', 'notes']
    const sets: string[] = []
    const values: any[] = []

    for (const field of fields) {
      if ((data as any)[field] !== undefined) {
        sets.push(`${field} = ?`)
        values.push((data as any)[field])
      }
    }

    // Auto-set resolved_date when status changes to resolved
    if (data.status === 'resolved' && !data.resolved_date) {
      const existing = await db.queryOne('SELECT resolved_date FROM disease_records WHERE id = ?', [id])
      if (existing && !existing.resolved_date) {
        sets.push('resolved_date = ?')
        values.push(now().split('T')[0])
      }
    }

    sets.push('updated_at = ?')
    values.push(now())
    values.push(id)

    await db.execute(`UPDATE disease_records SET ${sets.join(', ')} WHERE id = ?`, values)

    const row = await db.queryOne('SELECT * FROM disease_records WHERE id = ?', [id])
    return rowToRecord(row)
  },

  delete: async (id: string): Promise<void> => {
    await db.execute('DELETE FROM disease_records WHERE id = ?', [id])
  },

  addTreatment: async (diseaseId: string, data: DiseaseTreatmentCreate): Promise<DiseaseTreatment> => {
    const id = generateId()
    const userId = getLocalUserId()
    const timestamp = now()

    await db.execute(
      `INSERT INTO disease_treatments (id, disease_record_id, user_id, consumable_id, treatment_type,
       treatment_name, dosage, quantity_used, quantity_unit, treatment_date, duration_days,
       effectiveness, notes, created_at)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      [id, diseaseId, userId, data.consumable_id ?? null, data.treatment_type,
       data.treatment_name, data.dosage ?? null, data.quantity_used ?? null,
       data.quantity_unit ?? null, data.treatment_date, data.duration_days ?? null,
       data.effectiveness ?? null, data.notes ?? null, timestamp]
    )

    const row = await db.queryOne('SELECT * FROM disease_treatments WHERE id = ?', [id])
    return rowToTreatment(row)
  },

  updateTreatment: async (diseaseId: string, treatmentId: string, data: DiseaseTreatmentCreate): Promise<DiseaseTreatment> => {
    const fields = ['treatment_type', 'treatment_name', 'consumable_id', 'dosage',
      'quantity_used', 'quantity_unit', 'treatment_date', 'duration_days',
      'effectiveness', 'notes']
    const sets: string[] = []
    const values: any[] = []

    for (const field of fields) {
      if ((data as any)[field] !== undefined) {
        sets.push(`${field} = ?`)
        values.push((data as any)[field])
      }
    }
    values.push(treatmentId)
    values.push(diseaseId)

    await db.execute(
      `UPDATE disease_treatments SET ${sets.join(', ')} WHERE id = ? AND disease_record_id = ?`,
      values
    )

    const row = await db.queryOne('SELECT * FROM disease_treatments WHERE id = ?', [treatmentId])
    return rowToTreatment(row)
  },

  deleteTreatment: async (diseaseId: string, treatmentId: string): Promise<void> => {
    await db.execute(
      'DELETE FROM disease_treatments WHERE id = ? AND disease_record_id = ?',
      [treatmentId, diseaseId]
    )
  },

  getSummary: async (tankId: string): Promise<DiseaseHealthSummary> => {
    const userId = getLocalUserId()

    const activeCount = (await db.queryOne(
      'SELECT COUNT(*) as c FROM disease_records WHERE tank_id = ? AND user_id = ? AND status = ?',
      [tankId, userId, 'active']
    ))?.c || 0

    const monitoringCount = (await db.queryOne(
      'SELECT COUNT(*) as c FROM disease_records WHERE tank_id = ? AND user_id = ? AND status = ?',
      [tankId, userId, 'monitoring']
    ))?.c || 0

    const chronicCount = (await db.queryOne(
      'SELECT COUNT(*) as c FROM disease_records WHERE tank_id = ? AND user_id = ? AND status = ?',
      [tankId, userId, 'chronic']
    ))?.c || 0

    const resolvedCount = (await db.queryOne(
      'SELECT COUNT(*) as c FROM disease_records WHERE tank_id = ? AND user_id = ? AND status = ?',
      [tankId, userId, 'resolved']
    ))?.c || 0

    const totalTreatments = (await db.queryOne(
      `SELECT COUNT(*) as c FROM disease_treatments dt
       JOIN disease_records dr ON dt.disease_record_id = dr.id
       WHERE dr.tank_id = ? AND dr.user_id = ?`,
      [tankId, userId]
    ))?.c || 0

    const recentRows = await db.query(
      'SELECT * FROM disease_records WHERE tank_id = ? AND user_id = ? ORDER BY detected_date DESC LIMIT 5',
      [tankId, userId]
    )

    return {
      tank_id: tankId,
      active_count: activeCount,
      monitoring_count: monitoringCount,
      chronic_count: chronicCount,
      resolved_count: resolvedCount,
      total_treatments: totalTreatments,
      recent_diseases: recentRows.map(rowToRecord),
    }
  },
}

File: backend/app/api/v1/diseases.py
"""
Disease/Health Tracking API Endpoints

Handles disease records and treatment logs for livestock.
"""
from typing import List, Optional
from uuid import UUID
from datetime import date
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from sqlalchemy import func

from app.database import get_db
from app.models.user import User
from app.models.tank import Tank
from app.models.livestock import Livestock
from app.models.disease import DiseaseRecord, DiseaseTreatment
from app.models.consumable import Consumable, ConsumableUsage
from app.schemas.disease import (
    DiseaseRecordCreate,
    DiseaseRecordUpdate,
    DiseaseRecordResponse,
    DiseaseRecordDetailResponse,
    DiseaseTreatmentCreate,
    DiseaseTreatmentResponse,
    DiseaseHealthSummary,
)
from app.api.deps import get_current_user

router = APIRouter()


# ============================================================================
# Disease Records
# ============================================================================

@router.post("/", response_model=DiseaseRecordResponse, status_code=status.HTTP_201_CREATED)
def create_disease(
    disease_in: DiseaseRecordCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Create a new disease record."""
    # Verify livestock ownership
    livestock = db.query(Livestock).filter(
        Livestock.id == disease_in.livestock_id,
        Livestock.user_id == current_user.id
    ).first()
    if not livestock:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Livestock not found")

    # Verify tank ownership
    tank = db.query(Tank).filter(
        Tank.id == disease_in.tank_id,
        Tank.user_id == current_user.id
    ).first()
    if not tank:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Tank not found")

    data = disease_in.model_dump()
    disease = DiseaseRecord(**data, user_id=current_user.id)
    db.add(disease)
    db.commit()
    db.refresh(disease)
    return disease


@router.get("/", response_model=List[DiseaseRecordResponse])
def list_diseases(
    tank_id: Optional[UUID] = Query(None, description="Filter by tank ID"),
    livestock_id: Optional[UUID] = Query(None, description="Filter by livestock ID"),
    disease_status: Optional[str] = Query(None, alias="status", description="Filter by status"),
    severity: Optional[str] = Query(None, description="Filter by severity"),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """List disease records with optional filtering."""
    query = db.query(DiseaseRecord).filter(
        DiseaseRecord.user_id == current_user.id
    )

    if tank_id:
        tank = db.query(Tank).filter(Tank.id == tank_id, Tank.user_id == current_user.id).first()
        if not tank:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Tank not found")
        query = query.filter(DiseaseRecord.tank_id == tank_id)

    if livestock_id:
        query = query.filter(DiseaseRecord.livestock_id == livestock_id)

    if disease_status:
        query = query.filter(DiseaseRecord.status == disease_status)

    if severity:
        query = query.filter(DiseaseRecord.severity == severity)

    return query.order_by(DiseaseRecord.detected_date.desc()).all()


@router.get("/summary", response_model=DiseaseHealthSummary)
def get_health_summary(
    tank_id: UUID = Query(..., description="Tank ID"),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Get health summary for a tank."""
    tank = db.query(Tank).filter(Tank.id == tank_id, Tank.user_id == current_user.id).first()
    if not tank:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Tank not found")

    base = db.query(DiseaseRecord).filter(
        DiseaseRecord.tank_id == tank_id,
        DiseaseRecord.user_id == current_user.id
    )

    active_count = base.filter(DiseaseRecord.status == "active").count()
    monitoring_count = base.filter(DiseaseRecord.status == "monitoring").count()
    chronic_count = base.filter(DiseaseRecord.status == "chronic").count()
    resolved_count = base.filter(DiseaseRecord.status == "resolved").count()

    total_treatments = db.query(func.count(DiseaseTreatment.id)).join(DiseaseRecord).filter(
        DiseaseRecord.tank_id == tank_id,
        DiseaseRecord.user_id == current_user.id
    ).scalar() or 0

    recent = base.order_by(DiseaseRecord.detected_date.desc()).limit(5).all()

    return DiseaseHealthSummary(
        tank_id=tank_id,
        active_count=active_count,
        monitoring_count=monitoring_count,
        chronic_count=chronic_count,
        resolved_count=resolved_count,
        total_treatments=total_treatments,
        recent_diseases=recent,
    )


@router.get("/{disease_id}", response_model=DiseaseRecordDetailResponse)
def get_disease(
    disease_id: UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Get a disease record with its treatments."""
    disease = db.query(DiseaseRecord).filter(
        DiseaseRecord.id == disease_id,
        DiseaseRecord.user_id == current_user.id
    ).first()
    if not disease:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Disease record not found")
    return disease


@router.put("/{disease_id}", response_model=DiseaseRecordResponse)
def update_disease(
    disease_id: UUID,
    disease_in: DiseaseRecordUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Update a disease record."""
    disease = db.query(DiseaseRecord).filter(
        DiseaseRecord.id == disease_id,
        DiseaseRecord.user_id == current_user.id
    ).first()
    if not disease:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Disease record not found")

    update_data = disease_in.model_dump(exclude_unset=True)

    # Auto-set resolved_date when status changes to resolved
    if update_data.get("status") == "resolved" and not disease.resolved_date and not update_data.get("resolved_date"):
        update_data["resolved_date"] = date.today()

    for field, value in update_data.items():
        setattr(disease, field, value)

    db.commit()
    db.refresh(disease)
    return disease


@router.delete("/{disease_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_disease(
    disease_id: UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Delete a disease record and all its treatments."""
    disease = db.query(DiseaseRecord).filter(
        DiseaseRecord.id == disease_id,
        DiseaseRecord.user_id == current_user.id
    ).first()
    if not disease:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Disease record not found")
    db.delete(disease)
    db.commit()
    return None


# ============================================================================
# Disease Treatments
# ============================================================================

@router.post("/{disease_id}/treatments", response_model=DiseaseTreatmentResponse, status_code=status.HTTP_201_CREATED)
def add_treatment(
    disease_id: UUID,
    treatment_in: DiseaseTreatmentCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Add a treatment to a disease record."""
    disease = db.query(DiseaseRecord).filter(
        DiseaseRecord.id == disease_id,
        DiseaseRecord.user_id == current_user.id
    ).first()
    if not disease:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Disease record not found")

    # Verify consumable ownership if provided
    if treatment_in.consumable_id:
        consumable = db.query(Consumable).filter(
            Consumable.id == treatment_in.consumable_id,
            Consumable.user_id == current_user.id
        ).first()
        if not consumable:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Consumable not found")

    data = treatment_in.model_dump()
    treatment = DiseaseTreatment(**data, disease_record_id=disease_id, user_id=current_user.id)
    db.add(treatment)

    # Deduct from consumable stock if linked
    if treatment_in.consumable_id and treatment_in.quantity_used:
        consumable = db.query(Consumable).filter(
            Consumable.id == treatment_in.consumable_id
        ).first()
        if consumable:
            usage = ConsumableUsage(
                consumable_id=consumable.id,
                user_id=current_user.id,
                usage_date=treatment_in.treatment_date,
                quantity_used=treatment_in.quantity_used,
                quantity_unit=treatment_in.quantity_unit,
                notes=f"Treatment: {treatment_in.treatment_name}",
            )
            db.add(usage)

            if consumable.quantity_on_hand is not None:
                consumable.quantity_on_hand = max(0, consumable.quantity_on_hand - treatment_in.quantity_used)
                if consumable.quantity_on_hand <= 0:
                    consumable.status = "depleted"
                elif consumable.quantity_on_hand < treatment_in.quantity_used * 3:
                    consumable.status = "low_stock"

    db.commit()
    db.refresh(treatment)
    return treatment


@router.put("/{disease_id}/treatments/{treatment_id}", response_model=DiseaseTreatmentResponse)
def update_treatment(
    disease_id: UUID,
    treatment_id: UUID,
    treatment_in: DiseaseTreatmentCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Update a treatment."""
    treatment = db.query(DiseaseTreatment).filter(
        DiseaseTreatment.id == treatment_id,
        DiseaseTreatment.disease_record_id == disease_id,
        DiseaseTreatment.user_id == current_user.id
    ).first()
    if not treatment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Treatment not found")

    update_data = treatment_in.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(treatment, field, value)

    db.commit()
    db.refresh(treatment)
    return treatment


@router.delete("/{disease_id}/treatments/{treatment_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_treatment(
    disease_id: UUID,
    treatment_id: UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Delete a treatment."""
    treatment = db.query(DiseaseTreatment).filter(
        DiseaseTreatment.id == treatment_id,
        DiseaseTreatment.disease_record_id == disease_id,
        DiseaseTreatment.user_id == current_user.id
    ).first()
    if not treatment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Treatment not found")
    db.delete(treatment)
    db.commit()
    return None
